### PR TITLE
Turn on additional musttail checks for swiftc and runtime.

### PR DESCRIPTION
**Pull request by varungandhi-apple**
Fixes rdar://76024810.

### CODE REVIEW

File: unittests/runtime/CMakeLists.txt [hunk 1]
```diff
@@ -83,6 +83,13 @@ if(("${SWIFT_HOST_VARIANT_SDK}" STREQUAL "${SWIFT_PRIMARY_VARIANT_SDK}") AND
     endif()
   endif()
 
+  # Check that the tests are using the async calling convention
+  # correctly, and that musttail isn't missed by Clang/LLVM.
+  if(NOT SWIFT_COMPILER_IS_MSVC_LIKE)
+    set(CMAKE_CXX_FLAGS
+          "${CMAKE_CXX_FLAGS} -mllvm -enable-swifttailcc-musttail-check")
```
Comment (edymtt) [inline unittests/runtime/CMakeLists.txt:90]: A more idiomatic way to set these flags for `SwiftRuntimeTests` is to use `target_compile_options` and move this `if` before the finale `endif` in the file

```suggestion
    target_compile_options(SwiftRuntimeTests
      PUBLIC "-mllvm;-enable-swifttailcc-musttail-check")
```

This way we would match how the other properties for `SwiftRuntimeTests` are set after its definition, and we would suggest that those flags are not meant to be set globally for all targets.

File: include/swift/Option/Options.td [hunk 2]
```diff
@@ -1161,6 +1161,12 @@ def working_directory : Separate<["-"], "working-directory">,
 def working_directory_EQ : Joined<["-"], "working-directory=">,
   Alias<working_directory>;
 
+// LLVM verification
+def skip_swifttailcc_musttail_check : Flag<["-"], "skip-swifttailcc-musttail-check">,
+  Flags<[HelpHidden, DoesNotAffectIncrementalBuild]>,
+  HelpText<"Skip additional LLVM verification that all tail calls from "
+           "swifttailcc->swiftailcc are marked musttail.">;
```
Comment (harlanhaskins) [inline include/swift/Option/Options.td:1168]: nit: the second swifttailcc is misspelled

File: unittests/runtime/CMakeLists.txt (hunk 1 above)
Comment (varungandhi-apple) [reply unittests/runtime/CMakeLists.txt]: Fixed.

**Comment by varungandhi-apple** [comment]
@swift-ci test

**Comment by varungandhi-apple** [comment]
There is a legitimate verification failure:

```
define hidden swifttailcc void @"$s4main19_replacement_numberSiyYaFTI"(%swift.context* swiftasync %0) #7 {
  %2 = call i8* @swift_getOrigOfReplaceable50(i8** getelementptr inbounds (%swift.dyn_repl_link_entry, %swift.dyn_repl_link_entry* @"$s4main19_replacement_numberSiyYaFTX", i32 0, i32 0)) #1
  %3 = bitcast i8* %2 to void (%swift.context*)*
  %4 = bitcast void (%swift.context*)* %3 to %swift.async_func_pointer*
  %5 = getelementptr inbounds %swift.async_func_pointer, %swift.async_func_pointer* %4, i32 0, i32 0
  %6 = load i32, i32* %5, align 8
  %7 = sext i32 %6 to i64
  %8 = ptrtoint i32* %5 to i64
  %9 = add i64 %8, %7
  %10 = inttoptr i64 %9 to i8*
  %11 = bitcast i8* %10 to void (%swift.context*)*
  call swifttailcc void %11(%swift.context* swiftasync %0) // should be "musttail call"
  ret void
}
```

Need to fix that first before landing this.

**Comment by varungandhi-apple** [comment]
@swift-ci test

**Comment by varungandhi-apple** [comment]
@swift-ci test

**Comment by varungandhi-apple** [comment]
@swift-ci smoke test Windows

**Review by edymtt — APPROVED** [review]

**Comment by edymtt** [comment]
Approved with regard to CMake changes.

**Comment by davezarzycki** [comment]
Hi @varungandhi-apple – This change broke building Swift with top-of-tree clang as the host compiler and SWIFT_BUILD_RUNTIME_WITH_HOST_COMPILER set to TRUE. This regression hasn't happened in a very long time. Is there an ETA for getting this new compiler flag pushed to upstream LLVM?

CC: @rjmccall

**Comment by varungandhi-apple** [comment]
I did not know that such a configuration was supported.

I can submit a LLVM patch sometime today.

**Comment by davezarzycki** [comment]
Supported or not, that build option helps keep people honest about keeping divergences between upstream and downstream LLVM to a minimum. Thanks for looking into this. :-)

**Comment by varungandhi-apple** [comment]
LLVM patch: https://reviews.llvm.org/D103275

**Comment by davezarzycki** [comment]
Thanks. I know pushing changes upstream can be time consuming, but it is appreciated. Thanks!

**Comment by LucianoPAlmeida** [comment]
Hey @varungandhi-apple I was trying a clean Xcode build with `./swift/utils/build-script --xcode --release-debuginfo --debug-swift --skip-build-benchmarks  --swift-darwin-supported-archs 'x86_64' --sccache  --skip-ios --skip-tvos --skip-watchos --swift-disable-dead-stripping --skip-test-cmark`
And hit this error 
```
clang (LLVM option parsing)
: Unknown command line argument '-enable-swifttailcc-musttail-check'.  Try: 'clang (LLVM option parsing) --help'
clang (LLVM option parsing): Did you mean '--enable-implicit-null-checks'?
${MY_PATH}/swift-lang/build/Xcode-RelWithDebInfoAssert+swift-DebugAssert/swift-macosx-x86_64/stdlib/public/runtime/Swift.build/Debug/swiftRuntime-macosx-x86_64.build/Objects-normal/x86_64/SwiftDtoa.dia:1:1: warning: Could not read serialized diagnostics file: error("Failed to open diagnostics file") (in target 'swiftRuntime-macosx-x86_64' from project 'Swift')
Command CompileC failed with a nonzero exit code

** BUILD FAILED **

```
So the question is since this and the PR on the driver is merged, should this work now? I was going to clean and test again but since after clean it take hours to build want to make sure that it is necessary... 
Ah and sorry for commenting here (with risk of polluting the PR thread) if is better I can open a thread on the forums =]

**Comment by varungandhi-apple** [comment]
I've filed https://bugs.swift.org/browse/SR-14714 for this issue. Landing changes upstream will take a little bit more time (and will likely take a different shape than this flag), and fixing the Xcode build requires changing it to use the locally built Clang instead of the toolchain Clang.

For now, I'm going to revert this change and try to reland it later in a way that it doesn't break existing workflows.

**Comment by varungandhi-apple** [comment]
Revert PR: https://github.com/apple/swift/pull/37752

**Comment by LucianoPAlmeida** [comment]
Thank you for looking into this @varungandhi-apple :)

**Pull request merged**